js/packages/teams-ai/src/OpenAIApiFactory.ts

```diff
@@ -0,0 +1,407 @@
+    OpenAIClientResponse
+} from './OpenAIClients';
+import { Plan, PredictedDoCommand } from './Planner';
+import { ConversationHistory } from './ConversationHistory';
+
+export interface OpenAIApiFactoryOptions extends AIApiFactoryOptions {
+    /**
+     * The type of endpoint to use.
+     */
+    completionType?: 'chat' | 'text';
+
+    /**
+     * OpenAI API key.
+     */
+    apiKey: string;
+
+    /**
+     * Optional. OpenAI organization.
+     */
+    organization?: string;
+
+    /**
+     * Optional. OpenAI endpoint.
+     */
+    endpoint?: string;
+
+    /**
+     * The default model to use.
+     * @summary
+     * Prompts can override this model.
+     */
+    defaultModel: string;
+
+    /**
+     * Optional. A flag indicating if the planner should use the 'system' role when calling OpenAI's
+     * chatCompletion API.
+     * @summary
+     * The planner current uses the 'user' role by default as this tends to generate more reliable
+     * instruction following. Defaults to false.
+     */
+    useSystemMessage?: boolean;
+
+    /**
+     * Optional. A flag indicating if the planner should log requests to the console.
+     * @summary
+     * Both the prompt text and the completion response will be logged to the console. For
+     * chatCompletion calls the outgoing messages array will also be logged.
+     * Defaults to false.
+     */
+    logRequests?: boolean;
+}
+
+/**
+ * @template TState Type of the turn state object being persisted.
+ * @template TOptions Type of the options object being used.
+ */
+export class OpenAIApiFactory<TState extends TurnState = DefaultTurnState> implements AIApiFactory<TState> {
+    private readonly _options: OpenAIApiFactoryOptions;
+    private readonly _client: OpenAIClient;
+
+    public constructor(options: OpenAIApiFactoryOptions) {
+        this._options = Object.assign(
+            {
+                logRequests: false,
+                oneSayPerTurn: false,
+                useSystemMessage: false
+            } as OpenAIApiFactoryOptions,
+            options
+        );
+        this._client = this.createClient(this._options);
+    }
+
+    public async completePrompt(
+        context: TurnContext,
+        state: TState,
+        prompt: PromptTemplate,
+        options: ConfiguredAIOptions<TState>
+    ): Promise<string> {
+        if (this._options.completionType === 'text') {
+            const promptRequest = this.createCompletionRequest(prompt);
+            const result = await this.createCompletion(promptRequest);
+            return result.data?.choices[0]?.text ?? '';
+        } else {
+            const temp = (state['temp']?.value ?? {}) as DefaultTempState;
+            const chatRequest = this.createChatCompletionRequest(state, prompt, temp.input, options);
+            const result = await this.createChatCompletion(chatRequest);
+            return result.data?.choices[0]?.message?.content ?? '';
+        }
+    }
+
+    public async generatePlan(
```

Comment:
Should this method be here? It seems like it's not directly related to orchestrating the AI service.

Reply:
`generatePlan` is making a call to the service, so imo it is related. Let me know if I'm missing something.

Reply:
`createPlan`, the private helper method, is the central SDK-specific code for making the `Plan`.  (`createPlan` is called inside `generatePlan`)